File: JavaScript/ExampleTests/common/ValidValuesTyped.js
import { ValidValuesString } from "./ValidValuesString.js";
import * as _json from "./json.js";

export class ValidValuesTyped {
  constructor(value = "", isValid = false, notes = "") {
    this.value = value;
    this.isValid = isValid;
    this.notes = notes;
  }

  static fromStringObj(s) {
    return new ValidValuesTyped(
      s.value,
      ["true","t","yes","y","1"].includes(s.isValid.toLowerCase()),
      s.notes
    );
  }

  toStringObj() {
    return new ValidValuesString(
      String(this.value),
      String(this.isValid),
      String(this.notes)
    );
  }

  static toStringList(list) {
    return list.map(t => t.toStringObj());
  }

  static fromStringList(list) {
    return list.map(s => ValidValuesTyped.fromStringObj(s));
  }

  toJsonValue() {
    return {
      value: this.value,
      isValid: this.isValid,
      notes: this.notes,
    };
  }

  toJSON() { return _json.stringify(this.toJsonValue()); }

  static fromJsonValue(m) {
    return new ValidValuesTyped(
      _json.asString(_json.require(m, "value"), "value"),
      _json.asBool(_json.require(m, "isValid"), "isValid"),
      _json.asString(_json.require(m, "notes"), "notes")
    );
  }

  static fromJSON(text) { return ValidValuesTyped.fromJsonValue(_json.parse(text)); }

  static toJSONList(items) {
    return _json.stringify(items.map((item) => item.toJsonValue()));
  }

  static fromJSONList(text) {
    const raw = _json.asArray(_json.parse(text), "ValidValuesTyped");
    return raw.map((e) => ValidValuesTyped.fromJsonValue(e));
  }

  toString() {
    return `Value=${this.value}, IsValid=${this.isValid}, Notes=${this.notes}`;
  }

  equals(other) {
    if (!(other instanceof ValidValuesTyped)) return false;
    return this.value === other.value
      && this.isValid === other.isValid
      && this.notes === other.notes;
  }
}

File: Rust/ExampleTests/common/validvalues_typed.rs
#![allow(dead_code, unused_imports, unused_variables)]

use super::json;
use super::validvalues_string::ValidValuesString;

#[derive(Debug, Clone, Default, PartialEq)]
pub struct ValidValuesTyped {
    pub value: String,
    pub isvalid: bool,
    pub notes: String,
}

impl ValidValuesTyped {
    pub fn from_str_struct(s: &ValidValuesString) -> Self {
        Self {
            value: s.value.clone(),
            isvalid: matches!(s.isvalid.to_lowercase().as_str(), "true" | "t" | "yes" | "y" | "1"),
            notes: s.notes.clone(),
        }
    }

    pub fn to_str_struct(&self) -> ValidValuesString {
        ValidValuesString {
            value: self.value.clone(),
            isvalid: self.isvalid.to_string(),
            notes: self.notes.clone(),
        }
    }

    pub fn to_string_list(list: &[ValidValuesTyped]) -> Vec<ValidValuesString> {
        list.iter().map(|t| t.to_str_struct()).collect()
    }

    pub fn from_string_list(list: &[ValidValuesString]) -> Vec<ValidValuesTyped> {
        list.iter().map(ValidValuesTyped::from_str_struct).collect()
    }

    pub fn to_json_value(&self) -> json::Value {
        json::Value::Object(vec![
            ("Value".to_string(), json::Value::Str(self.value.clone())),
            ("IsValid".to_string(), json::Value::Bool(self.isvalid)),
            ("Notes".to_string(), json::Value::Str(self.notes.clone())),
        ])
    }

    pub fn to_json(&self) -> String {
        json::write(&self.to_json_value())
    }

    pub fn from_json_value(v: &json::Value) -> json::JsonResult<Self> {
        Ok(Self {
            value: json::as_string(json::require(v, "Value")?, "Value")?,
            isvalid: json::as_bool(json::require(v, "IsValid")?, "IsValid")?,
            notes: json::as_string(json::require(v, "Notes")?, "Notes")?,
        })
    }

    pub fn from_json(text: &str) -> json::JsonResult<Self> {
        Self::from_json_value(&json::parse(text)?)
    }

    pub fn to_json_list(list: &[ValidValuesTyped]) -> String {
        json::write(&json::Value::Array(
            list.iter().map(|item| item.to_json_value()).collect(),
        ))
    }

    pub fn from_json_list(text: &str) -> json::JsonResult<Vec<ValidValuesTyped>> {
        let root = json::parse(text)?;
        let items = json::as_array(&root, "ValidValuesTyped")?;
        let mut result = Vec::with_capacity(items.len());
        for e in items {
            result.push(Self::from_json_value(e)?);
        }
        Ok(result)
    }
}
